### PR TITLE
add chip GD32F450 of GD32MCU

### PR DESCRIPTION
**Pull request by GD32-MCU**
## Summary

Add the chip GD32F450 of GD32MCU to Nuttx

Modify arch/arm/Kconfig
Add gd32f4 in arch/arm/src folder
Modify boards/Kconfig
Add gd32f4 in boards/arm folder

## Testing
GPIO，network and NXFFS filesystems have been tested

### CODE REVIEW

**Comment by acassis** [comment]
Hi @GD32-MCU please check out board config files, they seems outdated. Please run for each board and config:
```
$ ./tools/refresh.sh --silent yourboard:yourconfig
```

**Comment by xiaoxiang781216** [comment]
@GD32-MCU could you squash the change into one patch?

File: arch/arm/src/gd32f4/Kconfig [hunk 1]
```diff
@@ -0,0 +1,2500 @@
...
+
+endif # GD32F4_TLI
+
+menu "GD32F4 QEncoder Driver"
+	depends on SENSORS_QENCODER
+	depends on GD32F4_TIMER0 || GD32F4_TIMER1 || GD32F4_TIMER2 || GD32F4_TIMER3 || GD32F4_TIMER4 || GD32F4_TIMER7
+
+endmenu # GD32F4 QEncoder Driver
+
+# endif # ARCH_CHIP_GD32F4
```
Comment (pkarashchenko) [inline arch/arm/src/gd32f4/Kconfig:2500]:
```suggestion
# endif # ARCH_CHIP_GD32F4

```

File: arch/arm/src/gd32f4/chip.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+ ****************************************************************************/
+
+/* Provide the required number of peripheral interrupt vector definitions as
+ * well. The definition GD32_IRQ_NEXTINT simply comes from the chip-specific
+ * IRQ header file included by arch/gd32f4/irq.h.
+ */
+
+#define ARMV7M_PERIPHERAL_INTERRUPTS  GD32_IRQ_NEXTINT
+
+#endif /* __ARCH_ARM_SRC_GD32_CHIP_H */
```
Comment (pkarashchenko) [inline arch/arm/src/gd32f4/chip.h:57]:
```suggestion
#endif /* __ARCH_ARM_SRC_GD32F4_CHIP_H */
```

File: arch/arm/src/gd32f4/chip.h [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __ARCH_ARM_SRC_GD32_CHIP_H
+#define __ARCH_ARM_SRC_GD32_CHIP_H
```
Comment (pkarashchenko) [inline arch/arm/src/gd32f4/chip.h:22]:
```suggestion
#ifndef __ARCH_ARM_SRC_GD32F4_CHIP_H
#define __ARCH_ARM_SRC_GD32F4_CHIP_H
```

File: arch/arm/src/gd32f4/gd32f4xx.h [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __ARCH_ARM_SRC_GD32F4_GD32F4xx_H
+#define __ARCH_ARM_SRC_GD32F4_GD32F4xx_H
+
```
Comment (pkarashchenko) [inline arch/arm/src/gd32f4/gd32f4xx.h:23]:
```suggestion
#ifndef __ARCH_ARM_SRC_GD32F4_GD32F4XX_H
#define __ARCH_ARM_SRC_GD32F4_GD32F4XX_H

```

File: arch/arm/src/gd32f4/gd32f4xx.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+#include "gd32f4xx_gpio.h"
+#include "gd32f4xx_rcu.h"
+#include "gd32f4xx_usart.h"
+#include "gd32f4xx_lowputc.h"
+#include "gd32f4xx_fmc.h"
+#include "gd32f4xx_pmu.h"
+#include "gd32f4xx_spi.h"
+#include "gd32f4xx_syscfg.h"
+
+#endif /* __ARCH_ARM_SRC_GD32F4_GD32F4xx_H */
```
Comment (pkarashchenko) [inline arch/arm/src/gd32f4/gd32f4xx.h:50]:
```suggestion
#endif /* __ARCH_ARM_SRC_GD32F4_GD32F4XX_H */
```

File: arch/arm/src/gd32f4/gd32f4xx_allocateheap.c [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+ *     Padding for alignment
+ *     User .data region         Size determined at link time
+ *     User .bss region          Size determined at link time
+ *     Kernel heap               Size determined by
+ *                                CONFIG_MM_KERNEL_HEAPSIZE
+ *     User heap                 Extends to the end of SRAM
+ *
+ ****************************************************************************/
+
+void up_allocate_heap(FAR void **heap_start, size_t *heap_size)
```
Comment (pkarashchenko) [inline arch/arm/src/gd32f4/gd32f4xx_allocateheap.c:251]:
```suggestion
void up_allocate_heap(void **heap_start, size_t *heap_size)
```
here and other similar places

File: arch/arm/src/gd32f4/gd32f4xx_dma.c [hunk 7]
```diff
@@ -0,0 +1,1051 @@
...
+
+/****************************************************************************
+ * Name: gd32_dma_channel_get
+ *
+ * Description:
+ *   Get the g_dmachan table entry associated with a DMA controller and
+ *   a channel number
+ ****************************************************************************/
+
+static inline FAR struct gd32_dma_channel_s
```
Comment (pkarashchenko) [inline arch/arm/src/gd32f4/gd32f4xx_dma.c:248]:
```suggestion
static inline struct gd32_dma_channel_s
```

**Comment by GD32-MCU** [comment]
> @GD32-MCU could you squash the change into one patch?

The commits has been squashed into one,

File: arch/arm/src/gd32f4/gd32f4xx_fmc.h [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+  FMC_READY,                       /* the operation has been completed */
+  FMC_BUSY,                        /* the operation is in progress */
+  FMC_RDDERR,                      /* read D-bus protection error */
+  FMC_PGSERR,                      /* program sequence error */
+  FMC_PGMERR,                      /* program size not match error */
+  FMC_WPERR,                       /* erase/program protection error */
+  FMC_OPERR,                       /* operation error */
+  FMC_PGERR,                       /* program error */
+  FMC_TOERR                        /* timeout error */
+}gd32_fmc_state_enum;
```
Comment (pkarashchenko) [inline arch/arm/src/gd32f4/gd32f4xx_fmc.h:47]:
```suggestion
} gd32_fmc_state_enum;
```

File: arch/arm/src/gd32f4/hardware/gd32f4xx_spi.h [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+#define GD32_SPI2                  GD32_SPI2_BASE
+#define GD32_SPI3                  GD32_SPI3_BASE
+#define GD32_SPI4                  GD32_SPI4_BASE
+#define GD32_SPI5                  GD32_SPI5_BASE
+
+#define GD32_I2S1_ADD              GD32_I2S1_ADD_BASE
+#define GD32_I2S2_ADD              GD32_I2S2_ADD_BASE
+
+/* SPI registers definitions */
+#define GD32_SPI_CTL0(spix)        (spix+GD32_SPI_CTL0_OFFSET)      /* SPI control register 0 */
```
Comment (pkarashchenko) [inline arch/arm/src/gd32f4/hardware/gd32f4xx_spi.h:72]:
```suggestion
#define GD32_SPI_CTL0(spix)        ((spix)+GD32_SPI_CTL0_OFFSET)      /* SPI control register 0 */
```
and other similar places

File: arch/arm/src/gd32f4/gd32f4xx_allocateheap.c (hunk 6 above)
Comment (GD32-MCU) [reply arch/arm/src/gd32f4/gd32f4xx_allocateheap.c:251]: Should I remove the “FAR ” from all submissions?

Comment (xiaoxiang781216) [reply arch/arm/src/gd32f4/gd32f4xx_allocateheap.c:251]: Yes, FAR is for the common code to support 16bit MCU, but ARM is 32bit arch.

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by jerpelea — APPROVED** [review]

**Comment by jifengrgj365** [comment]
@GD32-MCU 
Hi, There are no iic,can,adc drivers in GD32 chip. These are very useful. So, do you have a plan to add these?

**Comment by xiaoxiang781216** [comment]
See here: https://github.com/apache/nuttx/pull/7884

**Comment by jifengrgj365** [comment]
@GD32-MCU,@xiaoxiang781216
Nice!Thank you very much! I will test it.
We also need SDIO and CAN drivers for GD32 chip. 
I hope you can finish these in the future.
I'm sure it will help many people.

**Pull request merged**